File: bzip2/collect.py
import csv
import os
import re

LOG_DIR = "logs"
OUTPUT_CSV = "data.csv"

filename_pattern = re.compile(
    r"^(?P<framework>\w+!_)_"
    r"(?P<operation_workload>[\w\.-]+)"
    r"_iter(?P<iteration>\d+)"
)

# 存储所有数据行的临时列表
rows = []

for filename in os.listdir(LOG_DIR):
    if not filename.endswith('.log'):
        continue
    
    match = filename_pattern.search(filename)
    if not match:
        continue
    
    framework = match.group('framework').upper()
    operation_workload = match.group('operation_workload')
    iteration = int(match.group('iteration'))  # 转换为整数
    
    # 处理特殊文件名格式
    if ".log_" in operation_workload:
        operation_workload = operation_workload.replace(".log_", "_")
    
    # 读取执行时间
    with open(os.path.join(LOG_DIR, filename), 'r') as f:
        lines = f.readlines()
        if not lines:
            continue
            
        last_line = lines[-1].strip()
        if "Execution time:" not in last_line:
            continue
        
        try:
            time = float(last_line.split()[-2])
        except (IndexError, ValueError):
            continue
    
    # 收集数据（保持列顺序：Framework, Workload, Iteration, Time）
    rows.append((
        framework,
        operation_workload.replace('_', ' ').title(),
        iteration,
        f"{time:.6f}"
    ))

# 自定义排序函数
def sort_key(row):
    """ 
    排序优先级：
    1. 框架名称（字母顺序）
    2. 工作负载名称（字母顺序） 
    3. 迭代次数（数字大小）
    """
    return (row[0], row[1], row[2])

# 执行多级排序
sorted_rows = sorted(rows, key=sort_key)

# 写入排序后的CSV
with open(OUTPUT_CSV, 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerow(['Framework', 'Workload', 'Iteration', 'ExecutionTime(s)'])
    
    for row in sorted_rows:
        # 将迭代次数转换回字符串写入（如果需要保持原始格式）
        modified_row = list(row)
        modified_row[2] = str(modified_row[2])  # 转换迭代次数为字符串
        writer.writerow(modified_row)

print(f"saving data to {OUTPUT_CSV}")

File: bencher/src/collect.rs
use csv::Writer;
use heck::ToTitleCase;
use regex::Regex;
use std::error::Error;
use std::fs::{self, File};
use std::io::{BufRead, BufReader};

#[derive(Debug)]
struct LogData {
    framework: String,
    workload: String,
    iteration: u32,
    time: String,
}

pub fn collect_logs(log_dir: &str, output_csv: &str) -> Result<(), Box<dyn Error>> {
    // 编译正则表达式
    let re = Regex::new(
        r"^(?P<framework>[a-zA-Z-]+)_(?P<operation_workload>[\w._-]*)iter(?P<iteration>\d+)",
    )?;

    let mut rows = Vec::new();

    // 遍历日志目录
    for entry in fs::read_dir(log_dir)? {
        let path = entry?.path();
        if path.extension().and_then(|s| s.to_str()) != Some("log") {
            continue;
        }

        // 解析文件名
        let filename = path.file_name().unwrap().to_string_lossy();
        let caps = match re.captures(&filename) {
            Some(c) => c,
            None => continue,
        };

        // 提取字段
        let framework = caps["framework"].to_string();
        let operation_workload = caps["operation_workload"].replace(".log_", "_");
        let iteration = caps["iteration"].parse::<u32>()?;

        // 读取执行时间
        let file = File::open(&path)?;
        let lines: Vec<_> = BufReader::new(file).lines().collect::<Result<_, _>>()?;
        let last_line = match lines.last() {
            Some(line) => line.trim(),
            None => continue,
        };

        if !last_line.contains("Execution time:") {
            continue;
        }

        let time_str = match last_line.split_whitespace().nth_back(1) {
            Some(t) => t,
            None => continue,
        };

        // 格式化数据
        let workload = operation_workload.replace('_', " ").to_title_case();

        rows.push(LogData {
            framework,
            workload,
            iteration,
            time: format!("{:.6}", time_str.parse::<f64>()?),
        });
    }

    // 按框架、负载、迭代排序
    rows.sort_by(|a, b| {
        a.framework
            .cmp(&b.framework)
            .then(a.workload.cmp(&b.workload))
            .then(a.iteration.cmp(&b.iteration))
    });

    // 写入CSV文件
    let mut wtr = Writer::from_path(output_csv)?;
    wtr.write_record(&["Framework", "Workload", "Iteration", "ExecutionTime(s)"])?;

    for row in rows {
        wtr.write_record(&[
            row.framework,
            row.workload,
            row.iteration.to_string(),
            row.time,
        ])?;
    }

    wtr.flush()?;
    println!("saving data to {}", output_csv);

    Ok(())
}
